### PR TITLE
[DataGrid] Server filters

### PR DESCRIPTION
**Pull request by dtassone**
Fix most #633 

- Add filter props
- Add server filter
- Add filter stories

### CODE REVIEW

**Comment by dtassone** [comment]
I will open another PR for the rest
 - Column refactoring -> there is an issue whereby column reset sometimes after applying filter or resizing the grid
 - Component customization
 - Docs and Tests

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import { IconButton } from '@material-ui/core';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx:1]:
```suggestion
import IconButton from '@material-ui/core/IconButton';
```

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+export const ColumnHeaderFilterIcon: React.FC<ColumnHeaderFilterIconProps> = (props) => {
+  const { counter } = props;
+  const apiRef = React.useContext(ApiContext);
+  const options = useGridSelector(apiRef, optionsSelector);
+  const preferencePanel = useGridSelector(apiRef, preferencePanelStateSelector);
+
+  const tooltipContentNode = React.useMemo(() => {
+    return <div>{counter} active filter(s)</div>;
+  }, [counter]);
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx:24]: Why not use a string? A component prevents to get the aria label, we would need to apply it manually if we provide a component.

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+  const cssClasses = React.useMemo(
+    () =>
+      classnames(
+        HEADER_CELL_CSS_CLASS,
+        showColumnRightBorder ? 'MuiDataGrid-withBorder' : '',
+        column.headerClassName,
+        column.headerAlign === 'center' && 'MuiDataGrid-colCellCenter',
+        column.headerAlign === 'right' && 'MuiDataGrid-colCellRight',
+        {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx:93]: ?
```suggestion
        column.headerClassName,
        column.headerAlign === 'center' && 'MuiDataGrid-colCellCenter',
        column.headerAlign === 'right' && 'MuiDataGrid-colCellRight',
        {
          'MuiDataGrid-withBorder': showColumnRightBorder,
```

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    const colHeaderParams: ColParams = {
+      field: column.field,
+      colDef: column,
+      colIndex,
+      api: apiRef!.current,
+    };
+    apiRef!.current.publishEvent(COLUMN_HEADER_CLICK, colHeaderParams);
+  }, [apiRef, colIndex, column]);
+
+  const cssClasses = React.useMemo(
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx:85]: Did you measure a difference in performance?

```suggestion
  const cssClasses = 
```

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderMenuIcon.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        size="small"
+        onClick={menuIconClick}
+      >
+        {icon}
+      </IconButton>
+    </div>
+  );
+};
+
+ColumnHeaderMenuIcon.displayName = 'ColumnHeaderMenuIcon';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderMenuIcon.tsx:50]:
```suggestion

```

File: packages/grid/_modules_/grid/models/gridOptions.tsx [hunk 6]
```diff
@@ -182,6 +189,12 @@ export interface GridOptions {
    * Set it to 'server' if you would like to handle sorting on the server-side.
    */
   sortingMode?: FeatureMode;
+  /**
+   * Filtering can be processed on the server or client-side.
+   * Set it to 'client' if you would like to handle Filtering on the client-side.
+   * Set it to 'server' if you would like to handle Filtering on the server-side.
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridOptions.tsx:195]:
```suggestion
   * Set it to 'client' if you would like to handle filtering on the client-side.
   * Set it to 'server' if you would like to handle filtering on the server-side.
```

File: packages/grid/_modules_/grid/components/toolbar/FilterToolbarButton.tsx [hunk 7]
```diff
@@ -18,30 +20,42 @@ export const FilterToolbarButton: React.FC<{}> = () => {
...
         {counter} active filter(s)
         <ul>
-          {activeFilters.map((item) => (
-            <li key={item.id}>
-              {lookup[item.columnField!].headerName || item.columnField} {item.operatorValue}{' '}
-              {item.value}
-            </li>
-          ))}
+          {activeFilters.map((item) => ({
+            ...(lookup[item.columnField!] && (
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/toolbar/FilterToolbarButton.tsx:37]: Should we use a filter instead?

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeadersItemCollection.tsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  return (
+    <React.Fragment>
+      <GridColumnHeaderMenu />
+      <PreferencesPanel />
+      {items}
+    </React.Fragment>
+  );
+};
+
+ColumnHeaderItemCollection.displayName = 'ColumnHeaderItemCollection';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeadersItemCollection.tsx:65]:
```suggestion
```

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  const iconButton = (
+    <IconButton onClick={toggleFilter} color="default" aria-label="Show Filters" size="small">
+      {filteredColumnIconElement}
+    </IconButton>
+  );
+
+  return (
+    <Tooltip title={tooltipContentNode} enterDelay={1000}>
+      <div className="MuiDataGrid-sortIconContainer">
+        <div>
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx:58]: A specific reason for using two divs?

File: packages/grid/_modules_/grid/components/toolbar/FilterToolbarButton.tsx (hunk 7 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/components/toolbar/FilterToolbarButton.tsx:37]: No it's more performant like this

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx (hunk 2 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx:24]: > A component prevents

An element, sorry, wrong terminology.

File: packages/grid/_modules_/grid/components/toolbar/FilterToolbarButton.tsx (hunk 7 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/FilterToolbarButton.tsx:37]: Ok, I have never tried :)

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx (hunk 9 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx:54]: Yes it keeps the alignment with the icon and the badge on numeric columns

**Review by oliviertassinari — APPROVED** [review]
I think that we will need to change the workflow, we keep bundling changes that could be done isolated (making quality reviews hard, I have only surfaced it), introducing new features without documentation (making it hard to design great APIs), fixing bugs without test cases (open the doors to regressions).

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+          {headerComponent || (
+            <ColumnHeaderTitle
+              label={column.headerName || column.field}
+              description={column.description}
+              columnWidth={width}
+            />
+          )}
+          {!isColumnNumeric && columnTitleIconButtons}
+        </div>
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx:157]:
```suggestion
```

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        {!isColumnNumeric &&
+          !disableColumnMenu &&
+          !column.disableColumnMenu &&
+          columnMenuIconButton}
+      </div>
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx:163]:
```suggestion
```

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+      <ColumnHeaderSeparator
+        resizable={!disableColumnResize && !!column.resizable}
+        resizing={isResizing}
+        {...separatorProps}
+      />
+    </div>
+  );
+};
+ColumnHeaderItem.displayName = 'ColumnHeaderItem';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx:172]:
```suggestion
```

File: packages/grid/_modules_/grid/components/menu/columnMenu/ColumnsMenuItem.tsx [hunk 13]
```diff
@@ -0,0 +1,26 @@
+import MenuItem from '@material-ui/core/MenuItem';
+import * as React from 'react';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/menu/columnMenu/ColumnsMenuItem.tsx:2]: Should we keep on this convention?

```suggestion
import * as React from 'react';
import MenuItem from '@material-ui/core/MenuItem';
```

File: packages/storybook/src/stories/grid-filter.stories.tsx [hunk 14]
```diff
@@ -1,5 +1,20 @@
...
+  LinkOperator,
+  PreferencePanelsValue,
+  RowModel,
+  useApiRef,
+  XGrid,
+} from '@material-ui/x-grid';
+import { useDemoData } from '@material-ui/x-grid-data-generator';
+import { action } from '@storybook/addon-actions';
+import { useState } from 'react';
 import * as React from 'react';
```
Comment (oliviertassinari) [inline packages/storybook/src/stories/grid-filter.stories.tsx:15]:
```suggestion
import * as React from 'react';
import Button from '@material-ui/core/Button';
import {
  ColDef,
  ColTypeDef,
  FilterModel,
  LinkOperator,
  PreferencePanelsValue,
  RowModel,
  useApiRef,
  XGrid,
} from '@material-ui/x-grid';
import { useDemoData } from '@material-ui/x-grid-data-generator';
import { action } from '@storybook/addon-actions';
```

File: packages/storybook/src/stories/grid-filter.stories.tsx [hunk 15]
```diff
@@ -12,6 +27,251 @@ export default {
...
+            },
+          }}
+        />
+      </div>
+    </React.Fragment>
+  );
+}
+export function ServerFilterViaProps() {
+  const demoServer = useDemoData({ dataSet: 'Commodity', rowLength: 100 });
+  const [rows, setRows] = useState<RowModel[]>(demoServer.data.rows);
```
Comment (oliviertassinari) [inline packages/storybook/src/stories/grid-filter.stories.tsx:170]:
```suggestion
  const [rows, setRows] = React.useState<RowModel[]>(demoServer.data.rows);
```

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx [hunk 16]
```diff
@@ -0,0 +1,65 @@
+import IconButton from '@material-ui/core/IconButton';
+import Badge from '@material-ui/core/Badge';
+import Tooltip from '@material-ui/core/Tooltip';
+import * as React from 'react';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx:4]:
```suggestion
import * as React from 'react';
import IconButton from '@material-ui/core/IconButton';
import Badge from '@material-ui/core/Badge';
import Tooltip from '@material-ui/core/Tooltip';
```

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+import { PreferencePanelsValue } from '../../hooks/features/preferencesPanel/preferencesPanelValue';
+import { useIcons } from '../../hooks/utils/useIcons';
+import { optionsSelector } from '../../hooks/utils/useOptionsProp';
+import { ApiContext } from '../api-context';
+
+export interface ColumnHeaderFilterIconProps {
+  counter?: number;
+}
+
+export const ColumnHeaderFilterIcon: React.FC<ColumnHeaderFilterIconProps> = (props) => {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderFilterIcon.tsx:16]: No `children` prop is supported

```suggestion
export const ColumnHeaderFilterIcon = (props: ColumnHeaderFilterIconProps) => {
```

**Review by oliviertassinari — CHANGES_REQUESTED** [review]

**Comment by oliviertassinari** [comment]
Oops, I had some other suggestions in my draft

**Comment by dtassone** [comment]
> Oops, I had some other suggestions in my draft

I will integrate them in my next PR.

**Pull request merged**